edsl.md

```diff
@@ -62,17 +66,117 @@ where `F1 : F2 : F3 : F4` is the (two's complement) byte-array representation of
+
+    syntax Int ::= #lenOfHead ( TypedArg ) [function]
+ // -------------------------------------------------
+    rule #lenOfHead(#uint160( _ )) => 32
+    rule #lenOfHead(#address( _ )) => 32
+    rule #lenOfHead(#uint256( _ )) => 32
+    rule #lenOfHead( #int128( _ )) => 32
+    rule #lenOfHead(#bytes32( _ )) => 32
+    rule #lenOfHead(   #bool( _ )) => 32
+    rule #lenOfHead( #bytes(_, _)) => 32
+
+    syntax Bool ::= #isStaticType ( TypedArg ) [function]
  // -----------------------------------------------------
-    rule #getData(#uint160( DATA )) => #padToWidth(32, #asByteStack(DATA))
-    rule #getData(#address( DATA )) => #padToWidth(32, #asByteStack(DATA))
-    rule #getData(#uint256( DATA )) => #padToWidth(32, #asByteStack(DATA))
+    rule #isStaticType(#uint160( _ )) => true
+    rule #isStaticType(#address( _ )) => true
+    rule #isStaticType(#uint256( _ )) => true
+    rule #isStaticType( #int128( _ )) => true
+    rule #isStaticType(#bytes32( _ )) => true
+    rule #isStaticType(   #bool( _ )) => true
+    rule #isStaticType( #bytes(_, _)) => false
+
+    syntax Int ::= #sizeOfDynamicType ( TypedArg ) [function]
+ // ---------------------------------------------------------
+    rule #sizeOfDynamicType(#bytes(N, _)) => 32 +Int #ceil32(N)
+
+    syntax WordStack ::= #enc ( TypedArg ) [function]
+ // -------------------------------------------------
+    // static Type
+    rule #enc(#uint160( DATA )) => #padToWidth(32, #asByteStack(#getValue(#uint160( DATA ))))
+    rule #enc(#address( DATA )) => #padToWidth(32, #asByteStack(#getValue(#address( DATA ))))
+    rule #enc(#uint256( DATA )) => #padToWidth(32, #asByteStack(#getValue(#uint256( DATA ))))
+    rule #enc( #int128( DATA )) => #padToWidth(32, #asByteStack(#getValue( #int128( DATA ))))
+    rule #enc(#bytes32( DATA )) => #padToWidth(32, #asByteStack(#getValue(#bytes32( DATA ))))
+    rule #enc(   #bool( DATA )) => #padToWidth(32, #asByteStack(#getValue(   #bool( DATA ))))
+
+    // dynamic Type
+    rule #enc( #bytes(N, DATA)) => #enc(#uint256(N)) ++ #padToWidth(#ceil32(N), #asByteStack(DATA))
+
+    syntax Int ::= #getValue ( TypedArg ) [function]
+ // ------------------------------------------------
+    rule #getValue(#uint160( DATA )) => DATA
+      requires minUInt160 <=Int DATA andBool DATA <=Int maxUInt160
+
+    rule #getValue(#address( DATA )) => DATA
+      requires minUInt160 <=Int DATA andBool DATA <=Int maxUInt160
+
+    rule #getValue(#uint256( DATA )) => DATA
+      requires minUInt256 <=Int DATA andBool DATA <=Int maxUInt256
+
+    rule #getValue( #int128( DATA )) => #signed(DATA)
+      requires minSInt128 <=Int DATA andBool DATA <=Int maxSInt128
+
+    rule #getValue(#bytes32( DATA )) => DATA
+      requires minUInt256 <=Int DATA andBool DATA <=Int maxUInt256
+
+    rule #getValue(   #bool( DATA )) => DATA
+      requires 0 <=Int DATA andBool DATA <=Int 1
+
+    syntax Int ::= #signed ( Int ) [function]
+ // -----------------------------------------
```

Comment:
I think this belongs in `data.md`. Probably the `/sWord` and `%sWord` can be defined more cleanly in terms of this function. Also, I would call this version `#unsigned`, and add the inverse operation called `#signed` as well. Reason is to agree with wasm spec on naming convention.